src/utils/ucc_lock_free_queue.h

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCC_LOCKFREE_QUEUE_H_
+#define UCC_LOCKFREE_QUEUE_H_
+
+#include "utils/ucc_spinlock.h"
+#include "utils/ucc_atomic.h"
+#include <string.h>
+
+/* This data structure is thread safe */
+
+// Number of elements in a single lock free pool - could be changed, but in tests performed great
+#define LINE_SIZE 8
+#define NUM_POOLS 2
+
+typedef struct ucc_lf_queue_elem {
+    uint8_t         was_progressed;
+    ucc_list_link_t locked_list_elem;
+} ucc_lf_queue_elem_t;
+
+typedef struct ucc_lf_queue {
+    ucc_spinlock_t       locked_queue_lock[NUM_POOLS];
+    ucc_lf_queue_elem_t *elements[NUM_POOLS][LINE_SIZE];
+    uint8_t              which_pool;
+    ucc_list_link_t      locked_queue[NUM_POOLS];
+} ucc_lf_queue_t; // TODO the struct isn't a queue because not maintaining order, maybe another name
+
+static inline void ucc_lf_queue_enqueue(ucc_lf_queue_t *     queue,
+                                        ucc_lf_queue_elem_t *elem)
+{
+
+    int which_pool = elem->was_progressed ^ (queue->which_pool & 1);
+    int i;
+    for (i = 0; i < LINE_SIZE; i++) {
+        if (ucc_atomic_bool_cswap64(
+                (uint64_t *)&(queue->elements[which_pool][i]), 0,
+                (uint64_t)elem)) {
+            return;
+        }
+    }
+
+    ucc_spin_lock(&queue->locked_queue_lock[which_pool]);
+    ucc_list_add_tail(&queue->locked_queue[which_pool],
+                      &elem->locked_list_elem);
+    ucc_spin_unlock(&queue->locked_queue_lock[which_pool]);
+}
+
+static inline ucc_lf_queue_elem_t *ucc_lf_queue_dequeue(ucc_lf_queue_t *queue,
+                                                        int is_first_call)
+{
+    // Save value in the beginning of the function
+    int curr_which_pool = queue->which_pool;
+    int which_pool      = curr_which_pool & 1; // turn from even/odd -> bool
+    int i;
+    ucc_lf_queue_elem_t *elem;
+    for (i = 0; i < LINE_SIZE; i++) {
+        elem = queue->elements[which_pool][i];
+        if (elem) {
+            if (ucc_atomic_bool_cswap64(
+                    (uint64_t *)&(queue->elements[which_pool][i]),
+                    (uint64_t)elem, 0)) {
+                elem->was_progressed = 1;
+                return elem;
+            }
+        }
+    }
+    elem = NULL;
+    ucc_spin_lock(&queue->locked_queue_lock[which_pool]);
```

Comment:
So, an element from the locked list can only be extracted when LINE_SIZE elements are placed in the alternative pool. Looks like there is a big imbalance in favor of pools there. this might lead to the stalls in the progress.. i don't know if we can improve that - and make the dequeue extracting elements with the same probability rate ?

Reply:
I'm not sure what you meant. the locked queue is a direct continuation of the LINE_SIZE pool. when inserting, you search for an empty spot from the beginning of LINE_SIZE till the end of locked_queue. when you pop, you search for elem from the beginning of LINE_SIZE to the end of locked queue. 
Of course, as we know, it will not maintain order, meaning - new elems that were inserted after some other elem, could be popped before it, doesn't matter if they are on LINE_SIZE or locked queue. But, we have a guarantee that no "old" elem will be popped before some new elem. meaning, if some 2 objects A & B are in pool and B was popped and reinserted, we can assure that B will never be popped again before A will.